### PR TITLE
File blobstore support (#69)

### PR DESCRIPTION
**Pull request by charles-dyfis-net**
This is a modernization of #204 (prev. #202, see also #69).

### CODE REVIEW

**Comment by fdr** [comment]
Does this code work well for you? Can you get the tests to pass?

**Comment by charles-dyfis-net** [comment]
Argh -- sorry 'bout that. Had the tests running cleanly, then made some fixes based on testing in production and broke the style checks.

**Comment by charles-dyfis-net** [comment]
...so, *locally*, I also had to patch the tox configuration to force an older azure-storage release. Seems rather out-of-scope, though, so I'm leaving that out of this commit.

**Comment by fdr** [comment]
Looks like you got build of the beast ;)

But yeah, qualitatively, happy with how the patch is working? I see you added tests which is great.

**Comment by charles-dyfis-net** [comment]
Not entirely happy, actually, but for reasons related to missing functionality I couldn't justify folding into this PR (at least, not without prior discussion/consent):

I'm trying to write to storage where the postgresql user does not have ownership permissions. The attempts to make files and directories owner-accessible only rather than owner and group thus doesn't work for me -- the ownership is actually coerced away from the user who writes the files, meaning that content created 0700 in the filesystem in question isn't readable by the user who created it (directories can't be cd'd into by the user who created them, etc).

So, I'm pondering an adequate workaround for that -- but since it's a bizarre corner case unlikely to impact the vast majority of users, folding any such workaround into this patch strikes me as an unreasonable expansion of scope.

**Comment by fdr** [comment]
What about a setuid shim/hook that chowns/chmods files after the fact?

**Comment by charles-dyfis-net** [comment]
chowning is effectively done automatically by the filesystem -- you create something with UID x, it's stored with UID y, because *everything* on the filesystem is always, unconditionally, chowned to UID y (with no user, including root, able to prevent or override same). Hence, if you just did a mkdir with 0700 permissions, you can't *even as an immediately following syscall* cd into that directory you created, make a new file, there, etc.

Following along after-the-fact is thus much too late -- our attempt to store files in a directory wal-e just created will have already failed by the time we get there. Unless you mean to hook that directly into wal-e's makedirs() operations or similar, but that seems like overkill -- I'd rather just put `?perms=` on the `file://` url to let the user override the `0700` default.

**Comment by fdr** [comment]
Sure, why not add umask support? Couldn't you even change the umask before
wal-e starts?

Don't setuid programs get to be whatever user one wants? in the most
advanced case the program can run as root and then do whatever it needs to
do.

File: wal_e/blobstore/file/calling_format.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        fp = open(self.path, 'rb')
+        contents = fp.read()
+        fp.close()
+        return contents
+
+    def set_contents_from_file(self, fp):
+        ensure_dir_exists(self.path)
+        f = open(self.path, 'wb')
+        while True:
+            piece = fp.read(1024)
```
Comment (fdr) [inline wal_e/blobstore/file/calling_format.py]: Is this buffer a bit small? Should `copyfileobj` be used instead?

Comment (charles-dyfis-net) [inline wal_e/blobstore/file/calling_format.py]: Good call. Definitely doesn't make any sense for a buffer smaller than typical OS block or page size, or to fail to take advantage of a pertinent standard-library utility.

Comment (fdr) [inline wal_e/blobstore/file/calling_format.py]: another thing is to perhaps use the `with` construction to cleanly avoid exceptions causing resource leaks.

**Comment by charles-dyfis-net** [comment]
BTW -- do you want the review-feedback changes as a separate commit, or should I fold them in?

**Comment by fdr** [comment]
Can you sue the `git commit --fixup` option so I can both review them and squash them systematically?

**Comment by charles-dyfis-net** [comment]
(switching subjects -- yes, I can change the umask before wal-e starts, but wal-e actually hardcodes `0700` in some places, and the inverse of the umask is AND-ed into *requested* permissions, so umask alone isn't a powerful enough tool: I need to provide a way to override the default requested permisisons, and do so *only* when interacting with the target store -- since when we create something inside the postgresql data directory, we want to be conservative about access rights there; expect a separate PR when I have something hashed out I'm happy with).

**Comment by fdr** [comment]
(Don't worry about any patches you do without `--fixup`, I'll squash them later, if I want)

**Comment by charles-dyfis-net** [comment]
Ahh. Since everything's already committed, then, I'll let you do the squash. (Thanks for the heads-up on `commit --fixup`; first time I'd run into it).

**Comment by fdr** [comment]
Hmm. Do you think wal-e's paranoia about such a umask is overkill? I.e. should it set more bits and let the umask take care of things?

**Comment by charles-dyfis-net** [comment]
I'm generally in favor of paranoia-by-default; configurable overrides for the minority in need of same strikes me as reasonable.

**Comment by fdr** [comment]
Alright. Maybe if someone else asks or you find it particularly useful a
WAL_E_FILEMODE config can be implemented.

**Comment by charles-dyfis-net** [comment]
Just to be sure you're not waiting on me here -- is there anything else I should be doing with this PR, or is it waiting for cycles upstream?

**Comment by fdr** [comment]
Oh, no I'm just being slow

**Comment by fdr** [comment]
I committed it! (I made some very small changes).

**Pull request closed without merging**